File: sort/quicksort/quicksort.c
#include <stdio.h>


void print(int a[], int size) {
    int i;
    for (i = 0; i < size; i ++) {
        printf("%d ", a[i]);
    }
    printf("\n");
}


// 1 2 9 8 3  7  4 0 1 6 5

// sort by partition and then recursively
// pick right as pivot
int partition(int a[], int left, int right) {
    printf("+++++++++++++++\n");
    printf("Before Partition: \n");
    printf("left: %d, right: %d\n", left, right);
    print(a + left, right - left + 1);
    
    int i = left;
    int j = left;
    int tmp;
    for (j = left; j < right; j ++) {
        if (a[j] > a[right]) {
            if (a[i] < a[right]) {
                i = j;
            }
        } else {
            if (a[i] > a[right] ) {
                tmp = a[i];
                a[i] = a[j];
                a[j] = tmp;
            }
            i ++;
        }
        
        printf("a[i]: %d, a[j]: %d, a[right]: %d\n", a[i], a[j], a[right]);
    }
    
    // switch between i and the pivot
    tmp = a[right];
    a[right] = a[i];
    a[i] = tmp;
    
    printf("After Partition: \n");
    printf("i: %d\n", i);
    print(a + left, right - left + 1);
    printf("-------------------\n");
    
    return i;
}

void quickSort(int a[], int left, int right) {
    if (right <= left) {
        printf("right: %d <= left: %d\n", right, left);
        return ;
    }
    
    int i = partition(a, left, right);
    quickSort(a, left, i - 1);
    quickSort(a, i + 1, right);
}

void test(int a[], int size) {
    int i;
    printf("Before Sort: \n");
    print(a, size);
    printf("\n");
    
    quickSort(a, 0, size - 1);
    
    printf("\nAfter Sort: \n");
    print(a, size);
}

void main() {
    int a[] = {9, 8, 7, 6, 5, 4, 3, 3, 2, 1, 0 };
    // 1 2 9 8 3  7  4 0 1 6 5
    int b[] = {1, 2, 9, 8, 3, 7, 4, 0, 1, 6, 5};
    
    int c[] = {5, 8, 9, 23, 67, 1, 3, 7, 31, 56};
    // 1 3 5 7 8 9 23 31 56 67
 
    
    test(a, sizeof(a)/sizeof(a[0]));
    //test(b, sizeof(b)/sizeof(b[0]));

    //test(c, sizeof(c)/sizeof(c[0]));
}

File: sort/quicksort/quicksort.go
package main
import (
    "fmt"
)

/*
复杂度 NlogN (n次比较 * logN)
1.
从数列中挑出一个元素，称为 “基准”（pivot）
2.
重新排序数列，所有元素比基准值小的摆放在基准前面，所有元素比基准值大的摆在基准的后面（相同的数可以到任一边）。在这个分区退出之后，该基准就处于数列的中间位置。这个称为分区（partition）操作；
3.
递归地（recursive）把小于基准值元素的子数列和大于基准值元素的子数列排序。
*/

func Partition(a []int, left, right int) int {
    //1. 选择最左边一个作为pivot
    // lo 指向小于pivot的的所有值中最右边的值
    current, lo := left + 1, left
    
    
    //2. 循环中, 只需要和lo + 1对调， 不需要和pivot对调
    // 只需要和小于pivot的所有值的最右边值的下一个交换即可
    for current <= right { //= 是因为也要和最右边的值比较
        if a[current] < a[left] {
            lo ++
            a[current], a[lo] = a[lo], a[current]
        }
        current ++
    }
    
    //3. 交换pivot和小于pivot的最右边的值
    a[left], a[lo] = a[lo], a[left]
    return lo
}

func QuickSort(a []int, left, right int) {
    if left >= right {
        return
    }
    pos := Partition(a, left, right)
    QuickSort(a, left, pos - 1)
    QuickSort(a, pos + 1, right)
}

func Print(a []int) {
    for i := 0; i < len(a); i ++ {
        fmt.Printf("%d ", a[i])
    }
    fmt.Printf("\n")
}
func main() {
    a := []int{10, 10, 9, 7, 8, 6, 5, 4, 3, 2, 0, 1, 0}
    Print(a)
    
    QuickSort(a, 0, len(a) - 1)
    fmt.Println("After Sort:")
    Print(a)
}